crypto/s2n_rsa.c

```diff
@@ -24,89 +23,56 @@
 #include "crypto/s2n_hash.h"
 #include "crypto/s2n_openssl.h"
 #include "crypto/s2n_rsa.h"
+#include "crypto/s2n_pkey.h"
 
+#include "utils/s2n_blob.h"
 #include "utils/s2n_random.h"
 #include "utils/s2n_safety.h"
-#include "utils/s2n_blob.h"
-
-int s2n_asn1der_to_rsa_public_key(struct s2n_rsa_public_key *key, struct s2n_blob *asn1der)
-{
-    uint8_t *cert_to_parse = asn1der->data;
-    X509 *cert = d2i_X509(NULL, (const unsigned char **)(void *)&cert_to_parse, asn1der->size);
-    if (cert == NULL) {
-        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
-    }
-    /* If cert parsing is successful, d2i_X509 increments *cert_to_parse to the byte following the parsed data */
-    uint32_t parsed_len = cert_to_parse - asn1der->data;
-
-    if (parsed_len != asn1der->size) {
-        X509_free(cert);
-        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
-    }
 
-    EVP_PKEY *public_key = X509_get_pubkey(cert);
-    X509_free(cert);
+const struct s2n_pkey_ctx rsa_key_ctx = {
+    &s2n_rsa_sign,
+    &s2n_rsa_verify,
+    &s2n_rsa_encrypt,
+    &s2n_rsa_decrypt,
+    &s2n_rsa_keys_match,
+    &s2n_rsa_key_free
+};
 
-    if (public_key == NULL) {
-        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
-    }
-
-    if (EVP_PKEY_base_id(public_key) != EVP_PKEY_RSA) {
-        EVP_PKEY_free(public_key);
-        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
-    }
-
-    key->rsa = EVP_PKEY_get1_RSA(public_key);
-    if (key->rsa == NULL) {
-        EVP_PKEY_free(public_key);
+int s2n_pkey_to_rsa_public_key(s2n_rsa_public_key *rsa_key, EVP_PKEY *pkey)
+{
+    RSA *rsa = EVP_PKEY_get1_RSA(pkey);
+    if (rsa == NULL) {
         S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
     }
-
-    EVP_PKEY_free(public_key);
-
+    
+    rsa_key->rsa = rsa;
     return 0;
 }
 
-int s2n_asn1der_to_rsa_private_key(struct s2n_rsa_private_key *key, struct s2n_blob *asn1der)
+int s2n_pkey_to_rsa_private_key(s2n_rsa_private_key *rsa_key, EVP_PKEY *pkey)
 {
-    uint8_t *cert_to_parse = asn1der->data;
 
-    EVP_PKEY *pkey = d2i_PrivateKey(EVP_PKEY_RSA, NULL, (const unsigned char **)(void *)&cert_to_parse, asn1der->size);
-    if (pkey == NULL) {
+    RSA *rsa = EVP_PKEY_get1_RSA(pkey);
+    if (rsa == NULL) {
         S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
     }
-    RSA *rsa_key = EVP_PKEY_get1_RSA(pkey);
-    EVP_PKEY_free(pkey);
-    if (rsa_key == NULL) {
-        S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
-    }
-
-    /* If cert parsing is successful, d2i_RSAPrivateKey increments *cert_to_parse to the byte following the parsed data */
-    uint32_t parsed_len = cert_to_parse - asn1der->data;
-    if (parsed_len != asn1der->size) {
-        S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
-    }
-
-    if (!RSA_check_key(rsa_key)) {
+    
+    if (!RSA_check_key(rsa)) {
```

Comment:
It would take a little more finessing to get the correct type for the function argument and to create a generic function to satisfy both use cases. `s2n_rsa_check_key` is called to verify that the certificate parsing/validation function correctly set up the `s2n_pkey` object (which contains an `RSA` object). `RSA_check_key` is called on an `RSA` object inside of `s2n_pkey_to_rsa_private_key` to determine if the `RSA` object is valid before assigning it to an `s2n_rsa_private_key` object.